File: Python/0702-m-search-in-a-sorted-array-of-unknown-size.py
"""
Binary Search
"""
class Solution:
    def search(self, reader, target):
        """
        :type reader: ArrayReader
        :type target: int
        :rtype: int
        """
        index = 1
        while reader.get(index - 1) < target:
            index *= 2
        
        start, end = index // 2, index - 1
        while start + 1 < end:
            mid = start + (end - start) // 2
            if reader.get(mid) < target:
                start = mid
            else:
                end = mid
        
        if reader.get(start) == target:
            return start
        if reader.get(end) == target:
            return end
        
        return -1

File: leetcode/Java/0702-m-search-in-a-sorted-array-of-unknown-size.java
/*
Binary Search
*/
class Solution {
    public int search(ArrayReader reader, int target) {
        int index = 1;
        while (reader.get(index - 1) < target) {
            index = index * 2;
        }

        int start = index / 2, end = index - 1;
        while (start + 1 < end) {
            int mid = start + (end - start) / 2;
            if (reader.get(mid) < target) {
                start = mid;
            } else {
                end = mid;
            }
        }

        if (reader.get(start) == target) {
            return start;
        }

        if (reader.get(end) == target) {
            return end;
        }

        return -1;
    }
}